### PR TITLE
Fix load_morpholgies to always resolve paths

### LINKED ISSUE #1046: `neurom.load_morphologies` does not resolve paths
Works:
```python
import neurom
pop = neurom.load_morphologies("tests/data/valid_set")
[m.name for m in pop]
```
Fails:
```python
import neurom
pop = neurom.load_morphologies("tests/data/valid_set")
cd ..
[m.name for m in pop]
```
with:
```python
RawDataError                              Traceback (most recent call last)
File ~/Projects/NEUROM/NeuroM/neurom/core/population.py:87, in Population._load_file(self, f)
     86 try:
---> 87     return neurom.load_morphology(f)
     88 except (NeuroMError, MorphioError) as e:

File ~/Projects/NEUROM/NeuroM/neurom/io/utils.py:168, in load_morphology(morph, reader)
    166     return Morphology(_get_file(morph, reader))
--> 168 return Morphology(morph, Path(morph).name)

File ~/Projects/NEUROM/NeuroM/neurom/core/morphology.py:516, in Morphology.__init__(self, filename, name)
    510 """Morphology constructor.
    511 
    512 Args:
    513     filename (str|Path): a filename or morphio.{mut}.Morphology object
    514     name (str): an optional morphology name
    515 """
--> 516 self._morphio_morph = morphio.mut.Morphology(filename).as_immutable()
    517 self.name = name if name else 'Morphology'

RawDataError: File: tests/data/valid_set/Neuron_slice.h5 does not exist.

The above exception was the direct cause of the following exception:

NeuroMError                               Traceback (most recent call last)
Input In [8], in <cell line: 1>()
----> 1 [m.name for m in pop]

Input In [8], in <listcomp>(.0)
----> 1 [m.name for m in pop]

File ~/Projects/NEUROM/NeuroM/neurom/core/population.py:98, in Population.__iter__(self)
     96 """Iterator to populations's morphs."""
     97 for f in self._files:
---> 98     m = self._load_file(f)
     99     if m is None:
    100         continue

File ~/Projects/NEUROM/NeuroM/neurom/core/population.py:92, in Population._load_file(self, f)
     90         L.info('Ignoring exception "%s" for file %s', e, f.name)
     91     else:
---> 92         raise NeuroMError('`load_morphologies` failed') from e
     93 return None

NeuroMError: `load_morphologies` failed

```
Reported by @lidakanari

### PR DESCRIPTION
**Pull request by eleftherioszisis**
**Issue:** If a relative morphology directory is passed into `neurom.load_morphologies`, a Population object is created with relative paths to the morphologies. If the directory is changed after the Population is created, the paths will not be valid anymore.
**Changes:** Population makes paths absolute. Note that it does not resolve symlinks.
Fixes #1046

### CODE REVIEW

File: tests/io/test_io_utils.py [hunk 1]
```diff
@@ -128,6 +129,35 @@ def test_load_morphologies():
...
+    # store the dir of executing the tests
+    current_dir = os.getcwd()
+
+    # get the tests dir
+    tests_dir = Path(__file__).parent.parent
+
+    # move there for this test
+    os.chdir(tests_dir)
+
+    # get dir to morphs relative to tets dir
```
Comment (adrien-berchet) [inline tests/io/test_io_utils.py:143]:
```suggestion
    # get dir to morphs relative to test dir
```

**Review by adrien-berchet — COMMENTED** [review]
I am wondering if the path resolution should not be located in the `Population` constructor? Because creating a new `Population` instance will still have the same issue if the user does not use the `load_morphologies` function.

**Comment by eleftherioszisis** [comment]
> I am wondering if the path resolution should not be located in the `Population` constructor? Because creating a new `Population` instance will still have the same issue if the user does not use the `load_morphologies` function.

I am not so much in favor for the Population to do io stuff, as it is not its job.

**Comment by eleftherioszisis** [comment]
> > I am not so much in favor for the Population to do io stuff, as it is not its job.
> 
> But it already does, isn't it? If the cache is disabled, only file paths are stored in the Population object.

Well, yeah but my issue is that it can take either files or morphologies as the first argument. However, for the resolution to be correct, it cannot be lazy, because the directory can be changed at any time after init, before calling the iterators. So for this to work on the population, I would need to add a line that checks all files if they are morphologies or paths and then resolve them as needed, before letting cache and the rest do their own thing.

I would be more willing to add an assertion in `Population._load_file` for the path to be absolute. However, this is limiting if someone wants to create intentionally a population from relative paths.

**Comment by adrien-berchet** [comment]
Yeah, I prefer the first solution (resolving before the cache and other stuffs) because it makes it more consistent. But I agree that it would be cleaner to have a constructor that only accept a list of morphologies so the paths can be processed elsewhere but this would induce breaking changes. So as you prefer :-)

**Comment by eleftherioszisis** [comment]
Unfortunately, I cannot make Population fully resolve the paths, because there are test cases that use symlinks to morphologies with different filename...

**Comment by adrien-berchet** [comment]
I don't get what's the issue with symlinks to morphologies with different filename? Is it because the tests check the file names?

**Comment by adrien-berchet** [comment]
Hmmm ok
Is it really a feature (having morphology names of the population equal to the unresolved file name) or is it just something we check to be sure the test passed (in this case we could just change how we check this test)?

**Comment by eleftherioszisis** [comment]
> Hmmm ok Is it really a feature (having morphology names of the population equal to the unresolved file name) or is it just something we check to be sure the test passed (in this case we could just change how we check this test)?

No idea. I find it strange, but maybe it was a use case in the past.
https://github.com/BlueBrain/NeuroM/tree/master/tests/data/valid_set

**Review by adrien-berchet — APPROVED** [review]

File: CHANGELOG.rst [hunk 2]
```diff
@@ -4,6 +4,8 @@ Changelog
 Version 4.0.0
 -------------
 
+- Make ``neurom.core.Population`` resolve paths. Symlinks are resolved up until the
+  parent directory. If the file is symlinked, then it will not be resolved.
```
Comment (adrien-berchet) [inline CHANGELOG.rst:8]: Maybe we could also add a note on this in the docstring of `Population`?

File: neurom/core/population.py [hunk 3]
```diff
@@ -37,6 +38,21 @@
...
+    """Resolve the files in the list."""
+
+    def resolve_path(path):
+        """Resolve only the parent of the file path.
+
+        Not resolving the filename allows to have symlinks to files with different filename, but
+        any symlink in the parent tree would be correctly resolved.
+        """
+        path = Path(path)
+        return path.parent.resolve() / path.name
```
[the remark this answers is not in the document]
Comment (GianlucaFicarelli) [reply neurom/core/population.py:51]: I would prefer `path.absolute()` (but it's documented only from python 3.11 as we discussed) or even `Path(os.path.abspath(path))`, but I'm not strongly against the other options, so as you prefer.

[the remark this answers is not in the document]
Comment (eleftherioszisis) [reply neurom/core/population.py]: `Path(os.path.abspath(path))` is a good compromise. I will go with that :D

**Pull request merged**